chain/store/name.go

```diff
@@ -1,63 +1,178 @@
+type namesCleanup map[string]struct{}
+
+func newNameInfo(registrant []byte, expiresAt uint32) *nameInfo {
+	return &nameInfo{registrant: registrant, expiresAt: expiresAt}
 }
 
-func (sdb *StateDB) updateName(registrant []byte, name string) error {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) Serialize(w io.Writer) error {
+	if err := serialization.WriteVarBytes(w, ni.registrant); err != nil {
+		return err
+	}
+	if err := serialization.WriteUint32(w, ni.expiresAt); err != nil {
+		return err
+	}
+	return nil
+}
 
-	err := sdb.trie.TryUpdate(append(NamePrefix, nameId...), registrant)
+func (ni *nameInfo) Deserialize(r io.Reader) error {
+	var err error
+	ni.registrant, err = serialization.ReadVarBytes(r)
 	if err != nil {
 		return err
 	}
+	ni.expiresAt, err = serialization.ReadUint32(r)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
-	return sdb.trie.TryUpdate(append(NameRegistrantPrefix, registrantId...), []byte(name))
+func (ni *nameInfo) Bytes() ([]byte, error) {
+	buff := bytes.NewBuffer(nil)
+	err := ni.Serialize(buff)
+	if err != nil {
+		return nil, err
+	}
+	return buff.Bytes(), nil
 }
 
-func (sdb *StateDB) setName(registrant []byte, name string) {
-	registrantId := getRegistrantId(registrant)
-	nameId := getNameId(name)
+func (ni *nameInfo) String() string {
+	return fmt.Sprintf("%064x %d", ni.registrant, ni.expiresAt)
+}
+
+func (ni *nameInfo) Empty() bool {
+	return len(ni.registrant) == 0 && ni.expiresAt == 0
+}
+
+func getRegistrantId(registrant []byte) string {
+	return hex.EncodeToString(registrant)
+}
+
+func getNameId(name string, registrant []byte, expiresAt uint32) ([]byte, error) {
+	nameId := bytes.NewBuffer(nil)
+	nameId.Write([]byte(name))
+
+	ni := newNameInfo(registrant, expiresAt)
+	niBytes, err := ni.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	registrantHash := sha256.Sum256(niBytes)
+	nameId.Write(registrantHash[:hashPrefixLength])
+	return nameId.Bytes(), nil
+}
+
+func (sdb *StateDB) getRegName(name string) (string, error) {
+	iter := trie.NewIterator(sdb.trie.NodeIterator(append(NamePrefix, name...)))
+	iter.Next()
+	data := iter.Value
+	if len(data) > 0 {
+		buff := bytes.NewBuffer(data)
+		ni := new(nameInfo)
+		if err := ni.Deserialize(buff); err != nil {
+			return "", err
+		}
+		return ni.String(), nil
+	}
 
-	sdb.names.Store(registrantId, name)
-	sdb.nameRegistrants.Store(nameId, registrant)
+	return "", errors.New("no name registration info found")
 }
 
-func (sdb *StateDB) deleteName(registrantId string) error {
-	name, err := sdb.trie.TryGet(append(NameRegistrantPrefix, registrantId...))
+func (sdb *StateDB) getNameInfo(nameId []byte) (*nameInfo, error) {
+	if v, ok := sdb.names.Load(string(nameId)); ok {
+		if ni, ok := v.(*nameInfo); ok {
+			return ni, nil
+		}
+	}
+
+	ni := new(nameInfo)
+	enc, err := sdb.trie.TryGet(append(NamePrefix, nameId...))
+	if err != nil {
+		return nil, err
+	}
+	if len(enc) > 0 {
+		ni := new(nameInfo)
```

Comment:
This should be `ni = new(nameInfo)` instead of `ni := new(nameInfo)`. Also, it's better to move `ni := new(nameInfo)` after `TryGet` line to reduce memory allocation in case of error